### PR TITLE
Augmentation tests refactor

### PR DESCRIPTION
**Pull request by FGeri**
### Description
This is a work in progress pull request to refactor the test suit of the augmentation modul.

### Status
**Work in progress**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

**Comment by FGeri** [comment]
@edgarriba @shijianjian What do you think? This is still work in progress, but at the CommonTests and TestColorJitterAlternative, one can see the concept.

File: test/augmentation/test_augmentation.py [hunk 1]
```diff
@@ -25,6 +25,334 @@
...
+# apply_transform implementálva van és batchsiye stimmel
+
+
+class CommonTests(BaseTester):
+
+    ############################################################################################################
+    # Attributes variables to set
+    ############################################################################################################
+    _augmentation_cls = None
+    _default_param_set = None
```
Comment (shijianjian) [inline test/augmentation/test_augmentation.py:53]: I think this part is not that intuitive given ```param_set```. I think here, we may define a dummy example like 
```_augmentation = None```.  If need to initialize, you may do ```type(_augmentation)(**params)```.

File: test/augmentation/test_augmentation.py [hunk 2]
```diff
@@ -25,6 +25,334 @@
...
+        # Can be instatiated
+        augmentation = self._create_augmentation_from_params(**params, return_transform=False)
+        assert issubclass(
+            type(augmentation), AugmentationBase2D), f"{type(augmentation)} is not a subclass of AugmentationBase2D"
+
+        # generate_parameters can be called and returns the correct amount of parameters
+        batch_shape = (4, 3, 5, 6)
+        generated_params = augmentation.generate_parameters(batch_shape)
+        assert isinstance(generated_params, dict)
+        generated_params["batch_prob"] = torch.tensor([True, True, True, True])
```
Comment (shijianjian) [inline test/augmentation/test_augmentation.py:139]: Just to mention that the probabilities generation has been changed to a routine to avoid unnessasary computations:
```
select_images => generate_param_for_selected_images => apply_transform => replace_the_original_images
```
Thus, you may simply do:
```
generated_params = augmentation.generate_parameters((4, *batch_shape[1:]))
```
Then
```
images[batch_prob] = augmentation(selected_images, generated_params)
```
if you want params for 4 images.

**Review by shijianjian — COMMENTED** [review]
This is a very good start. Just some ideas to discuss.

File: test/augmentation/test_augmentation.py [hunk 3]
```diff
@@ -25,6 +25,334 @@
...
+
+class TestColorJitterAlternative(CommonTests):
+
+    possible_params = {
+        "brightness": (1., (0., 2.), [0., 2.], torch.tensor(0.), torch.tensor((0., 2.))),
+        "contrast": (1., (0., 2.), [0., 2.], torch.tensor(0.), torch.tensor((0., 2.))),
+        "saturation": (1., (0., 2.), [0., 2.], torch.tensor(0.), torch.tensor((0., 2.))),
+        "hue": (0., (-0.5, 0.5), [-0.5, 0.5], torch.tensor(0.), torch.tensor((-0.5, 0.5))),
+    }
+    params = [{param_name: v} for param_name, possible_values in possible_params.items() for v in possible_values]
```
[the remark this answers is not in the document]
Comment (FGeri) [reply test/augmentation/test_augmentation.py:307]: Have to check, but I guess the self is not yet in the scope at the fixture, though a classmethod might work.

File: test/augmentation/test_augmentation.py [hunk 4]
```diff
@@ -25,6 +25,334 @@
...
+# shape changing not shape changing
+#  return_transform: bool = False, same_on_batch: bool = False, p: float = 0.5,
+#  p_batch: float = 1.
+# Smoke tests, lehet példányosítani, meg lehet hívni az abstract függvényeit stb
+# generate_parameters annyi paramot ad vissza amennyi a batch size
+# compute_transformation implementálva van és batchsiye stimmel
+# apply_transform implementálva van és batchsiye stimmel
+
+
+class CommonTests(BaseTester):
```
[the remark this answers is not in the document]
Comment (FGeri) [reply test/augmentation/test_augmentation.py:47]: It's only my personal preference, but I think too many levels of abstraction and hence deep hierarchy makes the code sometimes really hard to understand. I'd go for something like CommonTests2D and CommonTests3D etc. It's not easy to define an absolut shared set of tests because a lot of some exceptions, but I think the cleanest would be in those cases to just override the necessary test cases or skip the ones that doesn't apply.

File: test/augmentation/test_augmentation.py (hunk 1 above)
Comment (FGeri) [reply test/augmentation/test_augmentation.py:53]: The concept behind, is that we want to create the augmentation objects within the test implementations to enforce certain behaviours by default (like p=0.0). _default_param_set is for having a default not trivial param set, that can be used for this.
As for param_set the name is a bit misleading, it's meant for the smoke tests to check if they can be run with every theoretically allowed arguments, I guess if we use them for every test that might be an overkill.

File: test/augmentation/test_augmentation.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (shijianjian) [reply test/augmentation/test_augmentation.py:47]: The only difference here is that CommonTests2D is doing (BCHW) while CommonTests3D is doing (BCDHW). What I am thinking is to have a CommonTester that can perform 1) shape-agnostic testings like test_smoke, and 2) some abstracted testings without shape awareness like "output.shape == expected_output.shape".

Then the child classes will do things like 1) shape-specific testings like testing HW, CHW, BCHW for 2D while DHW, CDHW, BCDHW for 3D, and 2) dimension-specific testing like the 3x3 return matrix for 2D and 4x4 return matrix for 3D.

File: test/augmentation/test_augmentation.py (hunk 1 above)
Comment (shijianjian) [reply test/augmentation/test_augmentation.py:53]: I think it is better to avoid having two suites of params. I think the ```param_set``` shall be taken as including all the edge cases?

\cc @edgarriba

**Comment by FGeri** [comment]
@edgarriba Any feedback? I'll soon refactor and further implement the remaining tests based on the comments.

File: test/augmentation/test_augmentation.py [hunk 5]
```diff
@@ -25,6 +25,334 @@
...
+
+
+class TestColorJitterAlternative(CommonTests):
+
+    possible_params = {
+        "brightness": (1., (0., 2.), [0., 2.], torch.tensor(0.), torch.tensor((0., 2.))),
+        "contrast": (1., (0., 2.), [0., 2.], torch.tensor(0.), torch.tensor((0., 2.))),
+        "saturation": (1., (0., 2.), [0., 2.], torch.tensor(0.), torch.tensor((0., 2.))),
+        "hue": (0., (-0.5, 0.5), [-0.5, 0.5], torch.tensor(0.), torch.tensor((-0.5, 0.5))),
+    }
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:306]: I would maybe just leave default parameters to and let users to define possible parameters to each test

File: test/augmentation/test_augmentation.py [hunk 6]
```diff
@@ -25,6 +25,334 @@
...
+
+    _augmentation_cls = ColorJitter
+    _default_param_set = {"brightness": 0.3,
+                          "contrast": 0.3,
+                          "saturation": 0.3,
+                          "hue": 0.3, }
+
+    @pytest.fixture(params=params, scope="class")
+    def param_set(self, request):
+        return request.param
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:317]: what's the use for this `param_set` fixture ?

File: test/augmentation/test_augmentation.py [hunk 7]
```diff
@@ -25,6 +25,334 @@
...
+    # Fixtures
+    ############################################################################################################
+    @pytest.fixture(scope="class")
+    def param_set(self, request):
+        raise NotImplementedError("param_set must be overriden in subclasses")
+
+    ############################################################################################################
+    # Test cases
+    ############################################################################################################
+    def test_smoke(self, device, dtype, param_set):
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:65]: I'm thinking that both `device` and `dtype` should be members of BaseTest class so that we can relieve each test call signature and just use `self.device` or `self.dtype` inside the implementation functions

File: test/augmentation/test_augmentation.py [hunk 8]
```diff
@@ -25,6 +25,334 @@
...
+        expected_transformation_shape = torch.Size((batch_shape[0], 3, 3))
+        test_input = torch.ones(batch_shape, device=device, dtype=dtype)
+        transformation = augmentation.compute_transformation(test_input, generated_params)
+        assert transformation.shape == expected_transformation_shape
+
+        # apply_transform can be called and returns the correct batch sized output
+        output = augmentation.apply_transform(test_input, generated_params)
+        assert output.shape[0] == batch_shape[0]
+
+    def _test_smoke_call_implementation(self, device, dtype, **params):
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:153]: would be too heavy to call 4 times the same augmentation here ?  same for others

Comment (edgarriba) [reply test/augmentation/test_augmentation.py:153]: test time is already large

File: test/augmentation/test_augmentation.py [hunk 9]
```diff
@@ -25,6 +25,334 @@
...
+        # fmt: off
+        input_tensor = torch.tensor([[.1, .2, .3, .4],
+                                     [.5, .6, .7, .8],
+                                     [.9, .0, .1, .2]], device=device, dtype=dtype).repeat(3, 1, 1)
+        expected_output = torch.tensor([[[[0.6024, 0.7273, 0.8522, 0.9771],
+                                          [1.0000, 1.0000, 1.0000, 1.0000],
+                                          [1.0000, 0.4775, 0.6024, 0.7273]]]], device=device, dtype=dtype).repeat(1, 3, 1, 1)
+        expected_transformation = torch.tensor([[[1., 0., 0.],
+                                                 [0., 1., 0.],
+                                                 [0., 0., 1.]]], device=device, dtype=dtype)
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:349]: can we take here inputs to `_test_random_p_1_return_transform_implementation` with decorators to make it clean ? /cc @anguelos

**Review by edgarriba — COMMENTED** [review]
I like it. Only one concern: in some test implementation we call 4 times the augmentation which could make this impossible to scale

File: test/augmentation/test_augmentation.py [hunk 10]
```diff
@@ -25,6 +25,334 @@
...
+
+        output, final_transformation = augmentation(
+            (test_input, test_transform), params=generated_params, return_transform=True)
+        assert output.shape[0] == batch_shape[0]
+        assert final_transformation.shape == expected_transformation_shape
+        assert_allclose(final_transformation, transformation @ test_transform)
+
+    def _test_consistent_output_shape_implementation(self, device, dtype, input_shape, **params):
+
+        # p==0.0
```
[the remark this answers is not in the document]
Comment (FGeri) [reply test/augmentation/test_augmentation.py:206]: In theory, yes. The cropping functions might cause some issues, though.

File: test/augmentation/test_augmentation.py [hunk 11]
```diff
@@ -25,6 +25,334 @@
...
+
+        # p==1.0
+        augmentation = self._create_augmentation_from_params(**params, p=1.0)
+        expected_output_shape = torch.Size((1,) * (4 - len(input_shape)) + tuple(input_shape))
+        test_input = torch.rand(input_shape, device=device, dtype=dtype)
+        output = augmentation(test_input)
+        assert len(output.shape) == 4
+        assert output.shape == expected_output_shape
+
+    def _test_random_p_0_implementation(self, device, dtype, **params):
```
[the remark this answers is not in the document]
Comment (FGeri) [reply test/augmentation/test_augmentation.py:222]: We could, but what would be the benefit?

File: test/augmentation/test_augmentation.py (hunk 5 above)
Comment (FGeri) [reply test/augmentation/test_augmentation.py:306]: Regarding this possible_params the idea was that I already bumped into situations when a theoretically valid argument set resulted in an exception. We could test this by parametrising the smoke tests to try every supported value for the single params. This possible_params dict is to put it into some ordered structure, that can be followed.

File: test/augmentation/test_augmentation.py (hunk 6 above)
Comment (FGeri) [reply test/augmentation/test_augmentation.py:317]: This fixture is used in the smoke tests in the CommonTests, one has to override it to be able to reparametrise it I think. This way one doesn't have to override and call here the smoke tests or any other test already implemented in the CommonTests, just if it doesn't fully fit in, f.e. ColorJitter requires 3 channels, so doesn't work for  test_consistent_output_shape with input shape (4,5)

File: test/augmentation/test_augmentation.py [hunk 12]
```diff
@@ -25,6 +25,334 @@
...
+                          "hue": 0.3, }
+
+    @pytest.fixture(params=params, scope="class")
+    def param_set(self, request):
+        return request.param
+
+    @pytest.mark.parametrize("input_shape", [(3, 4, 5), (2, 3, 4, 5)])
+    def test_consistent_output_shape(self, device, dtype, input_shape):
+        self._test_consistent_output_shape_implementation(
+            device=device, dtype=dtype, input_shape=input_shape, **self._default_param_set)
```
[the remark this answers is not in the document]
Comment (FGeri) [reply test/augmentation/test_augmentation.py:322]: No we don't need to that's the point, it's overloaded here because in the common tests it's tested with (4, 5) input size which colorjitter doesn't support.

File: test/augmentation/test_augmentation.py [hunk 13]
```diff
@@ -25,6 +25,334 @@
...
+
+    def test_random_p_1(self, device, dtype):
+        torch.manual_seed(42)
+        # fmt: off
+        input_tensor = torch.tensor([[.1, .2, .3, .4],
+                                     [.5, .6, .7, .8],
+                                     [.9, .0, .1, .2]], device=device, dtype=dtype).repeat(3, 1, 1)
+        expected_output = torch.tensor([[[[0.6024, 0.7273, 0.8522, 0.9771],
+                                          [1.0000, 1.0000, 1.0000, 1.0000],
+                                          [1.0000, 0.4775, 0.6024, 0.7273]]]]).repeat(1, 3, 1, 1)
```
[the remark this answers is not in the document]
Comment (FGeri) [reply test/augmentation/test_augmentation.py:332]: Right, I forgot to.

File: test/augmentation/test_augmentation.py (hunk 9 above)
Comment (FGeri) [reply test/augmentation/test_augmentation.py:349]: Hmm I 'm not sure if I understand what you mean.

File: test/augmentation/test_augmentation.py (hunk 8 above)
Comment (FGeri) [reply test/augmentation/test_augmentation.py:153]: Do you mean, that I should merge the 3 smoke tests?

File: test/augmentation/test_augmentation.py (hunk 5 above)
Comment (shijianjian) [reply test/augmentation/test_augmentation.py:306]: I have fixed some in #714. I think we might need to have ```valid_input_params``` to test all combinations if possible.

Meanwhile, I am thinking of submiting another PR to add tests for ```K.augmentation.random_generators``` and ```K.augmentation.functionals```.

**Comment by edgarriba** [comment]
@FGeri @shijianjian see how torchvision generate tests
https://github.com/pytorch/vision/commit/4106dbb864ce94e5003ef888ba23c59cc9c5ce31#diff-b91f9cf240d6656aaa67507d11139f5b8be05e9789b412544b5d515075141c2dR469

**Comment by shijianjian** [comment]
One more thing here, I have been discussing with @edgarriba and we are thinking of using something like
```torch.set_default_tensor_type('torch.cuda.FloatTensor')```  to set the default device and dtype. Then we may get rid of the duplicated device and dtype for all the function. You may consider this as well while refactoring.

**Comment by FGeri** [comment]
@edgarriba @shijianjian Next interation, coud you have look on it?

**Review by shijianjian — COMMENTED** [review]
Just few comments.

File: test/augmentation/test_augmentation.py [hunk 14]
```diff
@@ -24,6 +25,787 @@
...
+        return request.param
+
+    @pytest.mark.parametrize("input_shape,expected_output_shape",
+                             [((4, 5), (1, 1, 2, 3)), ((3, 4, 5), (1, 3, 2, 3)), ((2, 3, 4, 5), (2, 3, 2, 3))])
+    def test_cardinality(self, input_shape, expected_output_shape):
+        self._test_cardinality_implementation(input_shape=input_shape,
+                                              expected_output_shape=expected_output_shape,
+                                              params={"size": (2, 3), "align_corners": True})
+
+    @pytest.mark.xfail(reason="size=(1,2) results in RuntimeError: solve_cpu: For batch 0: U(3,3) is zero, singular U.")
```
[the remark this answers is not in the document]
Comment (FGeri) [reply test/augmentation/test_augmentation.py:388]: Sure, but I think then that should be then highlighted in the docs and validated upon construction at least.

File: test/augmentation/test_augmentation.py [hunk 15]
```diff
@@ -24,6 +25,787 @@
...
+    def test_random_p_1(self):
+        torch.manual_seed(42)
+
+        input_tensor = torch.tensor([[[0.1, 0.2, 0.3, 0.4],
```
[the remark this answers is not in the document]
Comment (FGeri) [reply test/augmentation/test_augmentation.py:392]: I think the issue is, that depending on the augmentation different input tensors make sense, F.e. equalise doesn't do anything with small images etc.

File: test/augmentation/test_augmentation.py [hunk 16]
```diff
@@ -24,6 +25,787 @@
...
+
+        torch.manual_seed(42)
+        out_sequence, transform_sequence = augmentation_sequence(input_tensor)
+
+        assert out2.shape == out_sequence.shape
+        assert transform.shape == transform_sequence.shape
+        assert_allclose(out2, out_sequence)
+        assert_allclose(transform, transform_sequence)
+
+    def _test_inverse_coordinate_check_implementation(self, params):
```
[the remark this answers is not in the document]
Comment (FGeri) [reply test/augmentation/test_augmentation.py:255]: I think this test is really important since it makes sure, the returned transformation matrix corresponds to the applied transformation. Ofc it could still go wrong but for a sainty check I think it's quite good. I found 4-5 of the bugs fixed in previous PRs with this check.

File: test/augmentation/test_augmentation.py [hunk 17]
```diff
@@ -24,6 +25,787 @@
...
+
+        assert_allclose(output_values, input_values)
+
+    def _test_gradcheck_implementation(self, params):
+        input_tensor = torch.rand((3, 5, 5), device=self.device, dtype=self.dtype)  # 3 x 3
+        input_tensor = utils.tensor_to_gradcheck_var(input_tensor)  # to var
+        assert gradcheck(
+            self._create_augmentation_from_params(
+                **params,
+                p=1.,
```
[the remark this answers is not in the document]
Comment (FGeri) [reply test/augmentation/test_augmentation.py:290]: Sometimes it does sometimes it fails even depending on the parametrisation of the augmentation. But we want this to pass in every case, right?

File: test/augmentation/test_augmentation.py [hunk 18]
```diff
@@ -24,6 +25,787 @@
...
+        assert output.shape == torch.Size((1,) * (4 - len(input_shape)) + tuple(input_shape))
+
+        # p==1.0
+        augmentation = self._create_augmentation_from_params(**params, p=1.0)
+        test_input = torch.rand(input_shape, device=self.device, dtype=self.dtype)
+        output = augmentation(test_input)
+        assert len(output.shape) == 4
+        assert output.shape == expected_output_shape
+
+    def _test_random_p_0_implementation(self, params):
```
[the remark this answers is not in the document]
Comment (FGeri) [reply test/augmentation/test_augmentation.py:197]: Not sure, with 0.5 I think it only make sense for a batch of at least 2 images. But then we cannot guarantee that let's say the first images is always transformed but the second not for every augmentation.

[the remark this answers is not in the document]
Comment (shijianjian) [reply test/augmentation/test_augmentation.py:197]: I think we can set a manual seed for that. In our base class, the random will firstly happen on batch_prob generators. So the batch_prob can be fixed using the same seed.

https://github.com/kornia/kornia/blob/87a0ead264ce5fc97997071acb9fe2286d6c425c/kornia/augmentation/base.py#L72-L90

File: test/augmentation/test_augmentation.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (shijianjian) [reply test/augmentation/test_augmentation.py:290]: Yes. Not sure if we can guarantee that.

File: test/augmentation/test_augmentation.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (shijianjian) [reply test/augmentation/test_augmentation.py:392]: Sure. Let's keep it as it is for now.

File: test/augmentation/test_augmentation.py [hunk 19]
```diff
@@ -24,6 +25,787 @@
...
+    ############################################################################################################
+    # Test cases
+    ############################################################################################################
+    def test_smoke(self, param_set):
+        self._test_smoke_implementation(params=param_set)
+        self._test_smoke_call_implementation(params=param_set)
+        self._test_smoke_return_transform_implementation(params=param_set)
+
+    @pytest.mark.parametrize("input_shape,expected_output_shape",
+                             [((4, 5), (1, 1, 4, 5)), ((3, 4, 5), (1, 3, 4, 5)), ((2, 3, 4, 5), (2, 3, 4, 5))])
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:58]: @FGeri how would you do here with random grayscale or any future augmentation that modifies the number of channels ?

File: test/augmentation/test_augmentation.py [hunk 20]
```diff
@@ -24,6 +25,787 @@
...
+
+        augmentation = self._create_augmentation_from_params(**params, p=1.0, return_transform=True)
+        output, transform = augmentation(input_tensor)
+
+        if (transform == kornia.eye_like(3, transform)).all():
+            pytest.skip("Test not relevant for intensity augmentations.")
+
+        grid_y, grid_x = torch.meshgrid(
+            torch.tensor(range(output.shape[-2]), device=self.device),
+            torch.tensor(range(output.shape[-1]), device=self.device))
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:269]: @FGeri you have kornia create_meshgrid :)

File: kornia/testing/__init__.py [hunk 21]
```diff
@@ -2,7 +2,10 @@
 The testing package contains testing-specific utilities.
 """
 from abc import ABC, abstractmethod
+from itertools import product
+from copy import deepcopy
 
+import pytest
```
Comment (edgarriba) [inline kornia/testing/__init__.py:8]: if we are making BaseTester dependent to `pytest` then we should move BaseTester outside the library core.

File: test/augmentation/test_augmentation.py (hunk 19 above)
Comment (FGeri) [reply test/augmentation/test_augmentation.py:58]: Simply by overriding the test and reparametrizing it.

File: test/augmentation/test_augmentation.py (hunk 20 above)
Comment (FGeri) [reply test/augmentation/test_augmentation.py:269]: I should discover the other modules as well :D

File: kornia/testing/__init__.py (hunk 21 above)
Comment (FGeri) [reply kornia/testing/__init__.py:8]: @edgarriba How do you prefer this part of the refactor? I could move to `BaseTester` out of library core but since `BaseTester` is already used at quite a few places that would involve some refactor or I could move the fixture into the `CommonTests`?

Comment (edgarriba) [reply kornia/testing/__init__.py:8]: I think it's better to move the tester under ./test - I can do a quick patch this weekend.

**Comment by shijianjian** [comment]
@FGeri Do you mean to merge this PR now?

File: kornia/testing/__init__.py (hunk 21 above)
Comment (FGeri) [reply kornia/testing/__init__.py:8]: I tried to move it to ./test but it didn't really work due to importing issues, so I went for the second options instead.

**Comment by FGeri** [comment]
> @FGeri Do you mean to merge this PR now?
@shijianjian I guess most or the findings are fixed, so it could.

**Review by shijianjian — APPROVED** [review]

File: test/augmentation/test_augmentation.py [hunk 22]
```diff
@@ -24,6 +25,794 @@
...
+        assert output.shape[0] == batch_shape[0]
+        assert transformation.shape == expected_transformation_shape
+
+        output, final_transformation = augmentation(
+            (test_input, test_transform), params=generated_params, return_transform=True)
+        assert output.shape[0] == batch_shape[0]
+        assert final_transformation.shape == expected_transformation_shape
+        assert_allclose(final_transformation, transformation @ test_transform)
+
+        output, transformation = augmentation((test_input, test_transform), params=generated_params)
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:160]: what's the difference between this call and the previous one ?

Comment (edgarriba) [reply test/augmentation/test_augmentation.py:160]: I miss more comments about what's happening

File: test/augmentation/test_augmentation.py [hunk 23]
```diff
@@ -24,6 +25,794 @@
...
+        assert_allclose(final_transformation, transformation @ test_transform)
+
+    def _test_cardinality_implementation(self, input_shape, expected_output_shape, params):
+
+        # p==0.0
+        augmentation = self._create_augmentation_from_params(**params, p=0.0)
+        test_input = torch.rand(input_shape, device=self.device, dtype=self.dtype)
+        output = augmentation(test_input)
+        assert len(output.shape) == 4
+        assert output.shape == torch.Size((1,) * (4 - len(input_shape)) + tuple(input_shape))
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:195]: what's this supposed to be doing when prob is zero? `output_shape==input.shape` no ?

File: test/augmentation/test_augmentation.py [hunk 24]
```diff
@@ -24,6 +25,794 @@
...
+        assert (output == test_input).all()
+        assert transformation.shape == expected_transformation_shape
+        assert (transformation == kornia.eye_like(3, transformation)).all()
+
+    def _test_random_p_1_implementation(self, input_tensor, expected_output, params):
+        augmentation = self._create_augmentation_from_params(**params, p=1.0, return_transform=False)
+        output = augmentation(input_tensor.to(self.device).to(self.dtype))
+
+        # Output should match
+        assert output.shape == expected_output.shape
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:227]: no against, but might be redundant since it's the same as we test in cardinality, right ?

File: test/augmentation/test_augmentation.py [hunk 25]
```diff
@@ -24,6 +25,794 @@
...
+
+        assert out2.shape == out_sequence.shape
+        assert transform.shape == transform_sequence.shape
+        assert_allclose(out2, out_sequence)
+        assert_allclose(transform, transform_sequence)
+
+    def _test_inverse_coordinate_check_implementation(self, params):
+        torch.manual_seed(42)
+
+        input_tensor = torch.zeros((1, 3, 50, 100), device=self.device, dtype=self.dtype)
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:265]: can we use a small tensor here - why that big ?

File: test/augmentation/test_augmentation.py [hunk 26]
```diff
@@ -24,6 +25,794 @@
...
+        input_tensor[:, :, 20:30, 40:60] = 1.
+
+        augmentation = self._create_augmentation_from_params(**params, p=1.0, return_transform=True)
+        output, transform = augmentation(input_tensor)
+
+        if (transform == kornia.eye_like(3, transform)).all():
+            pytest.skip("Test not relevant for intensity augmentations.")
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:272]: I wouldn't skip - it's also a good sanity check that intensity returns identity

File: test/augmentation/test_augmentation.py [hunk 27]
```diff
@@ -24,6 +25,794 @@
...
+
+        indices = kornia.create_meshgrid(height=output.shape[-2],
+                                         width=output.shape[-1],
+                                         normalized_coordinates=False,
+                                         device=self.device)
+        output_indices = indices.reshape((1, -1, 2)).to(dtype=self.dtype)
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:278]: @shijianjian probably we could include dtype to create_meshgrid too

File: test/augmentation/test_augmentation.py [hunk 28]
```diff
@@ -24,6 +25,794 @@
...
+        input_indices = kornia.geometry.transform_points(transform.to(self.dtype).inverse(), output_indices)
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:279]: @shijianjian why is not dtype inferred ?

File: test/augmentation/test_augmentation.py [hunk 29]
```diff
@@ -24,6 +25,794 @@
...
+
+        output_indices = output_indices.round().long().squeeze(0)
+        input_indices = input_indices.round().long().squeeze(0)
+        output_values = output[0, 0, output_indices[:, 1], output_indices[:, 0]]
+        value_mask = output_values > 0.9999
+
+        output_values = output[0, :, output_indices[:, 1][value_mask], output_indices[:, 0][value_mask]]
+        input_values = input_tensor[0, :, input_indices[:, 1][value_mask], input_indices[:, 0][value_mask]]
+
+        assert_allclose(output_values, input_values)
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:289]: we could improve this test or add an extra assertion firt so that the meshgrid coordinates behaves as follows:
`assert_allclose(points, K.transform_potsin(transform.inverse(), K.transform_points(transform, points)))`

File: test/augmentation/test_augmentation.py [hunk 30]
```diff
@@ -24,6 +25,794 @@
...
+        parameters = {}
+        self._test_random_p_1_return_transform_implementation(
+            input_tensor=input_tensor,
+            expected_output=expected_output,
+            expected_transformation=expected_transformation,
+            params=parameters)
+
+    @pytest.mark.skip(reason="No special parameters to validate.")
+    def test_exception(self):
+        pass
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:600]: I would just let to pass. Skipping I guess is for given some constrains or conditions.

File: test/augmentation/test_augmentation.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (edgarriba) [reply test/augmentation/test_augmentation.py:35]: I would rename this class as `CommonAugmentationTests` since it's only consumed by the augmentation module operators.

**Review by edgarriba — COMMENTED** [review]
Good overall, I added few comments mostly general question. Besides, I feel or example we need good documentation about how to write from minimal to proper tests in case someone wants to contribute with a new augmentation (e https://github.com/kornia/kornia/pull/773). Basically I guess a improve the CONTRIBUTING guide, specific for augmentations. https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst#coding-standards. As extra comment, I feel annoying that someone that is new to the library e.g. in front of a new  augmentation have to declare param_set each time. I guess with clear documentation might help. Good job!

**Comment by edgarriba** [comment]
@FGeri forgot  to ask, did you test this in a GPU ?

File: test/augmentation/test_augmentation.py (hunk 27 above)
Comment (shijianjian) [reply test/augmentation/test_augmentation.py:278]: Probably having device and dtype for all the functions.

If device and dtype are not None, do the conversion. Otherwise, perform a device and dtype sanity check with: (_extract_device_dtype) https://github.com/kornia/kornia/blob/master/kornia/utils/helpers.py#L6

File: test/augmentation/test_augmentation.py (hunk 28 above)
Comment (shijianjian) [reply test/augmentation/test_augmentation.py:279]: I found we have lots of legacy code having a bad device and dtype conversion during the run. We need to have proper testing and cuda CI here.

Might need to make the device and dtype param required for all the testing (they are optional currently).
